update the Lyrics screen

VASANTH04032004 opened the pull request:
Lyrics Screen Update:
  - Updated the `LyricsScreen` to display lyrics based on user input.
  - Implemented JSON serialization using `JsonSerializable` for the `Lyrics` model.
  - Added error handling to manage cases where lyrics are not found.

New Lyrics Input Screen:
  - Created a new `LyricsInputScreen` that allows users to enter the artist's name and song title.
  - Added validation to ensure both fields are filled before navigating to the lyrics display.

- **Model Refactoring**:
  - Refactored `SunriseSunset` and `Lyrics` models to use `@JsonSerializable` for better JSON serialization and deserialization.
  - Implemented the necessary manual methods for serialization as fallback.

Dependencies and Code Generation:
  - Updated `pubspec.yaml` to include `json_annotation`, `build_runner`, and `json_serializable`.
  - Ran `build_runner` to generate the serialization code for both models.

Changes Include:
- `LyricsScreen` updated for better handling of user input and JSON data.
- Added a new screen, `LyricsInputScreen`, for inputting artist and title.
- Refactored models for serialization with `JsonSerializable`.

Screen Recording:

Uploading Screen_recording_20240822_191041.mp4…

VASANTH04032004 committed:
update the Lyrics screen and create a new screen for get the particular Lyrics. change the modules using JsonSerializable

AnaghaVB commented:
Please change the title of the PR

AnaghaVB commented on lib/models/ sunrise_sunset_model.dart:
```diff
@@ -1,13 +1,31 @@
+@JsonSerializable()
 class SunriseSunset {
   final String sunrise;
   final String sunset;
```
Make the variables nullable

AnaghaVB commented on lib/models/lyrics_model.dart:
```diff
@@ -1,11 +1,28 @@
+
+@JsonSerializable()
 class Lyrics {
   final String lyrics;
```
Should be nullable

AnaghaVB commented on lib/screens/lyrics_input_screen.dart:
```diff
@@ -0,0 +1,80 @@
+                      border: OutlineInputBorder(),
+                    ),
+                  ),
+                  SizedBox(height: 16),
```
Remove SizedBox

AnaghaVB commented on lib/screens/lyrics_input_screen.dart:
```diff
@@ -0,0 +1,80 @@
+            ),
+          ),
+
+          Spacer(),
```
Remove

AnaghaVB commented on lib/screens/sunrise_sunset_screen.dart:
```diff
@@ -1,14 +1,15 @@
 
 class SunriseSunsetScreenState extends State<SunriseSunsetScreen> {
-  late Future<SunriseSunset> sunriseSunset;
+  late Future<Resource<SunriseSunset>> sunriseSunset;
```
Is there any reason you have defined this Resource type variable?

VASANTH04032004 replied:
The `Resource` type is defined to handle both successful and error states in a unified way, making it easier to manage and display API responses and errors in the UI.

AnaghaVB commented on lib/services/ api_service.dart:
```diff
@@ -1,14 +1,66 @@
 
-    if (response.statusCode == 200) {
-      return json.decode(response.body);
+    if (result.data != null) {
```
Please handle all kinds of exceptions and errors.

AnaghaVB commented on lib/services/ api_service.dart:
```diff
@@ -1,14 +1,66 @@
+class SunriseSunsetService {
+  final ApiService _apiService = ApiService();
+
+  Future<Resource<SunriseSunset>> fetchSunriseSunset(double lat, double lng) async {
```
Return type should simply be Future<Resource>

VASANTH04032004 committed:
Refactor API services to handle exceptions comprehensively

- Updated `ApiService` to catch specific exceptions including `ClientException`, `FormatException`, and generic exceptions.
- Enhanced error handling in `LyricsService` and `SunriseSunsetService` to provide detailed error messages.
- Changed return type of `fetchLyrics` and `fetchSunriseSunset` methods to `Future<Resource<T>>` for more robust error management.

VASANTH04032004 committed:
Refactor UI and API service handling:

- Updated `ApiService` to use `Success` and `Failure` classes for response handling.
- Removed use of `Map<String, dynamic>` and adjusted `get` method to handle JSON parsing using provided `fromJson` function.
- Modified `SunriseSunsetScreen` and `LyricsScreen` to use updated API service methods and improved error handling.
- Refactored UI layout for the lyrics input screen:
  - Removed unnecessary `Expanded` widget.

AnaghaVB commented on lib/services/ api_service.dart:
```diff
@@ -3,33 +3,46 @@ import 'package:http/http.dart' as http;
 
 class ApiService {
-  Future<Resource<Map<String, dynamic>>> get(String url) async {
+  Future<Resource<Item>> fetchData<Item>(String url, Item Function(String) fromJson) async {
```
The return type should only be Future<Resource>

AnaghaVB commented on lib/services/ api_service.dart:
```diff
@@ -3,33 +3,46 @@ import 'package:http/http.dart' as http;
 
-  Resource.success(this.data) : error = null;
-  Resource.failure(this.error) : data = null;
+abstract class Resource<Item> {}
```
Why is this an abstract class? Is Item a class? If so, remove that and pass generic type T

AnaghaVB commented on lib/services/ api_service.dart:
```diff
@@ -3,33 +3,46 @@ import 'package:http/http.dart' as http;
+abstract class Resource<Item> {}
+
+
+class Success<Item> extends Resource<Item> {
```
Why are these in the ApiService?

AnaghaVB commented on lib/services/ api_service.dart:
```diff
@@ -3,33 +3,46 @@ import 'package:http/http.dart' as http;
     } on Exception catch (e) {
-      return Resource.failure('Exception: ${e.toString()}');
+      return Failure(errorMessage: 'Exception: ${e.toString()}');
     } catch (e) {
```
Add Error handling also

VASANTH04032004 committed:
Refactor API handling to improve error management and modularity

- Enhanced error handling in `ApiService` with specific handling for `ClientException`, `FormatException`, `Exception`, `Error`, and unknown errors.
- Separated `Resource` classes into a dedicated file for better modularity.
- Updated `LyricsService` and `SunriseSunsetService` to use the new `Resource` class for clean and consistent API responses.
